File: matlab/KF.m
classdef KF
    %KF カルマンフィルタクラス
    properties
        model % A,B,G,H,R,history
        x % 状態値
        P % 誤差共分散
        K % カルマンゲイン
        x_hist % 状態値x_{k|k}の履歴
        P_hist % 状態値の誤差共分散P_{k|k}の履歴
        xp_hist % 予測値x_{k|k-1}の履歴
        Pp_hist % 予測値の誤差共分散P_{k|k-1}の履歴
    end
    methods
        function kf=KF(x0,P0,varargin)
            %KF カルマンフィルタを初期化する。
            %    kf=KF(x0,P0): 初期の状態値をx0、誤差共分散をP0としてカルマンフィルタkfを初期化する。
            %    kf=KF(x0,P0,model): 初期の状態値をx0、誤差共分散をP0、モデルをmodelとしてカルマンフィルタkfを初期化する。
            kf.x=x0;
            kf.P=P0;
            if 1<=length(varargin)
                kf.model=varargin{1};
                if isfield(kf.model,'history') && kf.model.history
                    kf.x_hist=x0;
                    kf.P_hist=diag(P0);
                end
            end
        end
        function pred=predict(kf,varargin)
            %PREDICT 次の時刻の状態値を予測する。
            %    pred=kf.predict(): 空の操作量が与えた次の時刻の状態値pred.xを予測し、誤差共分散行列pred.Pを与える。
            %    pred=kf.predict(u): 操作量uを与えた次の時刻の状態値pred.xを予測し、誤差共分散行列pred.Pを与える。
            %    pred=kf.predict(u,model): 操作量uとモデルmodelを与えた次の時刻の状態値pred.xを予測し、誤差共分散行列pred.Pを与える。
            %
            %モデルmodelは、オーバーライドしたい行列だけ含めば良いものとする。
            if 1<=length(varargin)
                u=varargin{1};
            else
                u=kf.null_u();
            end
            if 2<=length(varargin)
                kf.model=overwritestruct(kf.model,varargin{2});
            end
            pred=struct(...
                'x',kf.model.A*kf.x+kf.model.B*u,...
                'P',kf.model.A*kf.P*kf.model.A'+kf.model.G*kf.model.Q*kf.model.G');
        end
        function [kf,x,P,K,pred]=update(kf,z,varargin)
            %UPDATE 状態値を更新する。
            %    [kf,x,P,K,pred]=kf.update(z): 観測値zを与えた現在の状態値を更新し、状態値x、誤差共分散行列P、カルマンゲインK、事前予測値predを返す。
            %    [kf,x,P,K,pred]=kf.update(z,pred): 事前予測値predと観測値zを与えた現在の状態値を更新し、状態値x、誤差共分散行列P、カルマンゲインK、事前予測値predを返す。
            %    [kf,x,P,K,pred]=kf.update(z,pred,model): モデルmodelと事前予測値pred、観測値zを与えた現在の状態値を更新し、状態値x、誤差共分散行列P、カルマンゲインK、事前予測値predを返す。
            %
            %モデルmodelは、オーバーライドしたい行列だけ含めば良いものとする。
            if 1<=length(varargin)
                pred=varargin{1};
            else
                pred=kf.predict();
            end
            if 2<=length(varargin)
                kf.model=overwritestruct(kf.model,varargin{2});
            end
            K=pred.P*kf.model.H'/(kf.model.R+kf.model.H*pred.P*kf.model.H');
            KH=K*kf.model.H;
            I=eye(size(KH));
            x=pred.x+K*(z-kf.model.H*pred.x);
            P=(I-KH)*pred.P;
            
            kf.x=x;
            kf.P=P;
            kf.K=K;
            if isfield(kf.model,'history') && kf.model.history
                kf.x_hist=[kf.x_hist x];
                kf.P_hist=[kf.P_hist diag(P)];
                kf.xp_hist=[kf.xp_hist pred.x];
                kf.Pp_hist=[kf.xp_hist diag(pred.P)];
            end
        end
        function u=null_u(kf)
            %NULL_U 空の入力を作成する。
            u=zeros(size(kf.model.B,2),1);
        end
        function [dist,conf]=eval_dist(kf,z,varargin)
            %EVAL_DIST 予測値と観測値の距離を評価する。
            %    [dist,conf]=kf.eval_dist(z): 予測値と観測値zの距離distと確信度（距離を標準偏差で割った値のノルム）confを得る。
            %    [dist,conf]=kf.eval_dist(z,pred): 予測値predを用いる。
            %    [dist,conf]=kf.eval_dist(z,pred,model): 予測値predとモデルmodelを用いる。
            %
            %観測値zは1個以上の縦ベクトルとする。
            %距離distと確信度confはそれぞれ、観測値zに対応する1個以上の縦ベクトルおよびスカラ値の配列とする。
            if 1<=length(varargin)
                pred=varargin{1};
            else
                pred=kf.predict();
            end
            if 2<=length(varargin)
                kf.model=overwrite(kf.model,varargin{2});
            end
            dist=z-kf.model.H*pred.x;
            devi=kf.model.H*sqrt(diag(pred.P));
            conf=vecnorm(dist./devi,2,1);
        end
    end
end

File: python/KF.py
import numpy
import numpy.linalg

class KF:
    def __init__(self,x0,P0,A=None,B=None,G=None,Q=None,H=None,R=None,history=True):
        """
        Constructs a KF object.

        Parameters
        ----------
        x0 : D x 1 numpy.array
            initial state, which must be a vertical vector
        P0 : D x D numpy.array
            initial error covariance matrix for x0
        A : D x D numpy.array
            the state transition matrix
        B : D x E numpy.array
            the control input model
        G : D x D numpy.array
            the process noise transition matrix
        Q : D x D numpy.array
            the covariance of the process noise
        H : C x D numpy.array
            the observation model
        R : C x C numpy.array
            the covariance of the observation noise
        history : bool
        """
        self.x=x0.reshape((x0.size,1))
        self.P=P0
        self.A=A
        self.B=B
        self.G=G
        self.Q=Q
        self.H=H
        self.R=R
        self.history=history
        if self.history:
            self.hist_x=self.x
            self.hist_P=numpy.transpose([numpy.diag(self.P)])
            self.hist_xp=numpy.empty((self.hist_x.shape[0],0))
            self.hist_Pp=numpy.empty((self.hist_P.shape[0],0))

    def predict(self,u=None,A=None,B=None,G=None,Q=None):
        """
        Predicts the state at one step ahead in time

        Parameters
        ----------
        u : E x 1 numpy.array
            control vector
        A : D x D numpy.array
            the state transition matrix
        B : D x E numpy.array
            the control input model
        G : D x D numpy.array
            the process noise transition matrix
        Q : D x D numpy.array
            the covariance of the process noise
        """
        if A is not None:
            self.A=A
        if B is not None:
            self.B=B
        if G is not None:
            self.G=G
        if self.G is None:
            self.G=numpy.eye(A.shape[0],A.shape[1])
        if Q is not None:
            self.Q=Q
        xp=numpy.dot(self.A,self.x.reshape((self.x.size,1)))
        if u is not None and self.B is not None:
            xp=xp+numpy.dot(self.B,u.reshape((u.size,1)))
        Pp=numpy.dot(numpy.dot(self.A,self.P),self.A.T)+numpy.dot(numpy.dot(self.G,self.Q),self.G.T)
        return (xp,Pp)

    def update(self,z,pred=None,H=None,R=None):
        """
        Updates the state with the given observation

        Parameters
        ----------
        z : C x 1 numpy.array
            observation
        H : C x D numpy.array
            the observation model
        R : C x C numpy.array
            the covariance of the observation noise
        """
        z=z.reshape((z.size,1))
        if pred is None:
            pred=self.predict()
        if H is not None:
            self.H=H
        if self.H is None:
            self.H = numpy.eye(z.shape[0],self.A.shape[0])
        if R is not None:
            self.R=R
        xp,Pp=pred[0],pred[1]

        K=numpy.dot(numpy.dot(Pp,self.H.T),numpy.linalg.inv(self.R+numpy.dot(numpy.dot(self.H,Pp),self.H.T)))
        KH=numpy.dot(K,self.H)
        x=xp+numpy.dot(K,(z-numpy.dot(self.H,xp)))
        P=numpy.dot(numpy.eye(KH.shape[0],KH.shape[1])-KH,Pp)

        self.x,self.P=x,P
        if self.history:
            self.hist_x=numpy.append(self.hist_x,self.x,axis=1)
            self.hist_P=numpy.append(self.hist_P,numpy.transpose([numpy.diag(self.P)]),axis=1)
            self.hist_xp=numpy.append(self.hist_xp,xp,axis=1)
            self.hist_Pp=numpy.append(self.hist_Pp,numpy.transpose([numpy.diag(Pp)]),axis=1)
        return (self.x,self.P)

    def eval_dist(self,z,pred=None,H=None):
        if z.shape[1] is None:
            dist=z.reshape((z.size,1))
        if pred is None:
            pred=self.predict()
        if H is None:
            H=self.H
        xp,Pp=pred[0],pred[1]
        dist=z.T
        Hxp=numpy.dot(H,xp)
        for e in dist:
            e[...]=e-Hxp[:,0]
        dist=dist.T
        devi=numpy.dot(H,numpy.sqrt(numpy.transpose(numpy.diag(Pp))))
        devi=devi.reshape((devi.size,1))
        conf=numpy.linalg.norm(dist/devi,ord=2,axis=0)
        conf=conf.reshape((1,conf.size))
        return dist,conf

if __name__ == "__main__":
    import numpy.random
    import matplotlib
    import matplotlib.pyplot as pyplot

    softsensor=True

    T=300
    dt=0.1

    #
    # DYNAMICS
    #

    A=numpy.array([
        [1, 0, dt, 0],
        [0, 1, 0, dt],
        [0, 0, 1, 0],
        [0, 0, 0, 1]])
    B=numpy.array([
        [dt*dt/2, 0],
        [0, dt*dt/2],
        [dt, 0],
        [0, dt]])
    G=numpy.eye(4)
    Q=0.1*numpy.eye(4)
    H=numpy.eye(4)
    if softsensor:
        R=3*numpy.array([
            [1, 0, 0, 0],
            [0, 1, 0, 0],
            [0, 0, 2/dt, 0],
            [0, 0, 0, 2/dt]])
    else:
        R=3*numpy.array([
            [1, 0, 0, 0],
            [0, 1, 0, 0],
            [0, 0, 1, 0],
            [0, 0, 0, 1]])

    #
    # GROUNDTRUETH GENERATION
    #

    vmean=numpy.zeros((4,1))
    wmean=numpy.zeros((4,1))
    v=vmean+numpy.dot(numpy.linalg.cholesky(Q),numpy.random.randn(A.shape[0],T))
    w=wmean+numpy.dot(numpy.linalg.cholesky(R),numpy.random.randn(H.shape[0],T))

    x=numpy.zeros((A.shape[0],T))
    x[:,0]=numpy.transpose(20*numpy.random.rand(A.shape[1],1))
    u=numpy.random.rand(B.shape[1],T)
    u[0.4<u]=0
    u=u-0.5
    z=numpy.zeros((H.shape[0],T))
    z[:,0]=numpy.dot(H,x[:,0])+w[:,0]
    if softsensor:
        z[2:4,0]=0 # indifferentiable at t=0
    for k in range(1,T):
        x[:,k]=numpy.dot(A,x[:,k-1])+numpy.dot(B,u[:,k-1])+v[:,k-1]
        z[:,k]=numpy.dot(H,x[:,k])+w[:,k]
        if softsensor:
            z[2:4,k]=(z[0:2,k]-z[0:2,k-1])/dt # overwrite velocity with differentiated positions

    #
    # FILTERING
    #

    kf=KF(z[:,0],R,A=A,B=B,G=G,Q=Q,H=H,R=R,history=True)
    for k in range(1,T):
        pred=kf.predict(u[:,k])
        kf.update(z[:,k],pred=pred)

    #
    # PLOT
    #

    # two components of state vector to be plot below
    comp1=0
    comp2=1
    pyplot.plot(x[comp1,:],x[comp2,:],label='$x_k$',color='black',linestyle='solid')
    pyplot.scatter(z[comp1,:],z[comp2,:],label='$z_k$',color='red')
    pyplot.plot(kf.hist_x[comp1,:],kf.hist_x[comp2,:],label='$x_{k|k}$',color='blue')
    pyplot.legend()
    pyplot.show()
